src/main/java/org/privacyidea/JSONParser.java

```diff
@@ -0,0 +1,386 @@
+            if (root != null) {
+                try {
+                    JsonObject obj = root.getAsJsonObject();
+                    return obj.getAsJsonObject(RESULT).getAsJsonObject(VALUE).getAsJsonPrimitive(TOKEN).getAsString();
+                } catch (Exception e) {
+                    privacyIDEA.error("Response did not contain an authorization token: " + formatJson(serverResponse));
+                }
+            }
+        } else {
+            privacyIDEA.error("/auth response was empty or null!");
+        }
+        return null;
+    }
+
+    /**
+     * Parse the response of the server into a PIResponse object.
+     *
+     * @param serverResponse response of the server
+     * @return PIResponse or null if input is empty
+     */
+    PIResponse parsePIResponse(String serverResponse) {
+        if (serverResponse == null || serverResponse.isEmpty()) return null;
+
+        PIResponse response = new PIResponse();
+        response.rawMessage = serverResponse;
+
+        JsonObject obj;
+        try {
+            obj = JsonParser.parseString(serverResponse).getAsJsonObject();
+        } catch (JsonSyntaxException e) {
+            privacyIDEA.error(e);
+            return response;
+        }
+
+        response.id = getString(obj, ID);
+        response.piVersion = getString(obj, VERSION_NUMBER);
+        response.signature = getString(obj, SIGNATURE);
+        response.jsonRPCVersion = getString(obj, JSONRPC);
+
+        JsonObject result = obj.getAsJsonObject(RESULT);
+        if (result != null) {
+            response.status = getBoolean(result, STATUS);
+            response.value = getBoolean(result, VALUE);
+
+            JsonElement errElem = result.get(ERROR);
+            if (errElem != null && !errElem.isJsonNull()) {
+                JsonObject errObj = result.getAsJsonObject(ERROR);
+                response.error = new PIResponse.Error();
+                response.error.code = getInt(errObj, CODE);
+                response.error.message = getString(errObj, MESSAGE);
+            }
+        }
+
+        JsonElement detailElem = obj.get(DETAIL);
+        if (detailElem != null && !detailElem.isJsonNull()) {
+            JsonObject detail = obj.getAsJsonObject(DETAIL);
+            response.message = getString(detail, MESSAGE);
+            response.serial = getString(detail, SERIAL);
+            response.transactionID = getString(detail, TRANSACTION_ID);
+            response.type = getString(detail, TYPE);
+            response.otpLength = getInt(detail, OTPLEN);
+
+            JsonArray arrMessages = detail.getAsJsonArray(MESSAGES);
+            if (arrMessages != null) {
+                arrMessages.forEach(val -> {
+                    if (val != null) {
+                        response.messages.add(val.getAsString());
+                    }
+                });
+            }
+
+            JsonArray arrChallenges = detail.getAsJsonArray(MULTI_CHALLENGE);
```

Comment:
First: can we do the parsing inside the corresponding classes?
Like adding a constructor which accepts a json string or object and takes the values it needs (Challenge, TokenInfo, PIResponse).

Second: in python we work a lot with exceptions. Can't we just use something like
```
try {
    obj.getJsonObject(DETAIL).getString(SERIAL);
    ...
} except Error {
    continue
}
```
like shown here: https://www.oracle.com/technical-resources/articles/java/json.html
This would eliminate a lot of checks for Null, empty or type.

Also it looks like the Gson library is primarily used to (de)serialize Java objects as Json but we mostly have some plain data. Maybe a different library like [`javax.json`](https://javadoc.io/static/javax.json/javax.json-api/1.1.4/index.html?overview-summary.html) is more straight forward?

Reply:
> 
> 
> First: can we do the parsing inside the corresponding classes?
> Like adding a constructor which accepts a json string or object and takes the values it needs (Challenge, TokenInfo, PIResponse).

It was implemented like that before. It has the problem that there is no privacyIDEA instance available in the constructor to log errors to (unless you pass one aswell which is also meh). Doing it this way, the parser class is kind of a "factory" and all json parsing is done in one place (which also eliminates the references to the helper methods).
> 
> Second: in python we work a lot with exceptions. Can't we just use something like
> 
> ```
> try {
>     obj.getJsonObject(DETAIL).getString(SERIAL);
>     ...
> } except Error {
>     continue
> }
> ```
> 
> like shown here: https://www.oracle.com/technical-resources/articles/java/json.html
> This would eliminate a lot of checks for Null, empty or type.
> 
> Also it looks like the Gson library is primarily used to (de)serialize Java objects as Json but we mostly have some plain data. Maybe a different library like [`javax.json`](https://javadoc.io/static/javax.json/javax.json-api/1.1.4/index.html?overview-summary.html) is more straight forward?

If an exception is thrown, the execution of the following statements is skipped. However, it is possible that, for example, the multi_challenge array is empty and the rest of the response should be parsed properly afterward. JSON parsing is pretty verbose in java unforunately unless it is done automatically with objects.

javax.json was used before and iirc it was annoying to manage as dependency because it is split into the api definition and the implementation (which is not part of javax), therefore i moved to Gson. Also i wanted to check if i could automate the serialization of the responses to objects but didnt find the time yet or forgot. The code is about the same no matter which json parser is used.